File: app/src/main/java/mb00/android/codehub/ui/repo/view/RepoIssuesFragment.java
package mb00.android.codehub.ui.repo.view;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import mb00.android.codehub.R;
import mb00.android.codehub.api.builder.RetrofitBuilder;
import mb00.android.codehub.api.service.GitHubService;
import mb00.android.codehub.data.BundleKeys;
import mb00.android.codehub.data.PreferenceKeys;
import mb00.android.codehub.databinding.FragmentRepoIssuesBinding;
import mb00.android.codehub.ui.base.view.BaseBindingFragment;
import mb00.android.codehub.ui.repo.adapter.IssueAdapter;
import mb00.android.codehub.ui.repo.adapter.RepoFragmentPagerAdapter;
import mb00.android.codehub.ui.repo.viewmodel.RepoIssuesViewModel;
import retrofit2.Retrofit;
import timber.log.Timber;

/**
 * Fragment containing repository issues; launched from {@link RepoFragmentPagerAdapter}
 */

public class RepoIssuesFragment extends BaseBindingFragment<FragmentRepoIssuesBinding, RepoIssuesViewModel> {

    private String authHeader;
    private String userName;
    private String repoName;

    @Override
    protected int layout() {
        return R.layout.fragment_repo_issues;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        SharedPreferences preferences = getActivity().getSharedPreferences(PreferenceKeys.PREFERENCES, Context.MODE_PRIVATE);
        authHeader = preferences.getString(PreferenceKeys.AUTH_HEADER, "");
        userName = getArguments().getString(BundleKeys.USER_NAME);
        repoName = getArguments().getString(BundleKeys.REPO_NAME);
    }

    @Override
    public void onStart() {
        super.onStart();

        initRecyclerView();

        getBinding().repoIssuesSwipeRefreshLayout.setOnRefreshListener(() -> {
            repoCodeCall(authHeader, userName, repoName);
            getBinding().repoIssuesSwipeRefreshLayout.setRefreshing(false);
        });
        repoCodeCall(authHeader, userName, repoName);
    }

    private void initRecyclerView() {
        getBinding().repoIssuesRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        getBinding().repoIssuesRecyclerView.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL));
    }

    private void repoCodeCall(String header, String user, String repo) {
        Retrofit retrofit = RetrofitBuilder.getInstance();
        GitHubService service = retrofit.create(GitHubService.class);

        service.getRepoIssues(header, user, repo)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(repoIssueList -> {
                    if (repoIssueList.size() > 0) {
                        IssueAdapter issueAdapter = new IssueAdapter(repoIssueList);
                        getBinding().repoIssuesRecyclerView.setAdapter(issueAdapter);
                    } else {
                        getBinding().noIssuesTextView.setVisibility(View.VISIBLE);
                    }
                }, error -> Timber.e(error.getMessage()));
    }

}

File: app/src/main/java/mb00/android/codehub/ui/repo/view/RepoIssuesFragment.kt
package mb00.android.codehub.ui.repo.view

import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import mb00.android.codehub.R
import mb00.android.codehub.data.BundleKeys
import mb00.android.codehub.data.PreferenceKeys
import mb00.android.codehub.databinding.FragmentRepoIssuesBinding
import mb00.android.codehub.ui.base.view.BaseBindingFragment
import mb00.android.codehub.ui.repo.adapter.IssueAdapter
import mb00.android.codehub.ui.repo.adapter.RepoFragmentPagerAdapter
import mb00.android.codehub.ui.repo.viewmodel.RepoIssuesViewModel
import timber.log.Timber

/**
 * Fragment containing repository issues; launched from [RepoFragmentPagerAdapter]
 */

class RepoIssuesFragment : BaseBindingFragment<FragmentRepoIssuesBinding, RepoIssuesViewModel>() {

    private lateinit var authHeader: String
    private lateinit var userName: String
    private lateinit var repoName: String

    override fun layout(): Int {
        return R.layout.fragment_repo_issues
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val preferences = activity?.getSharedPreferences(PreferenceKeys.PREFERENCES, Context.MODE_PRIVATE)
        authHeader = preferences?.getString(PreferenceKeys.AUTH_HEADER, "") ?: ""
        userName = arguments?.getString(BundleKeys.USER_NAME) ?: ""
        repoName = arguments?.getString(BundleKeys.REPO_NAME) ?: ""

    }

    override fun onStart() {
        super.onStart()

        initRecyclerView()
        initSwipeRefreshLayout()

        repoIssuesCall(authHeader, userName, repoName)
    }

    private fun initRecyclerView() {
        binding.repoIssuesRecyclerView.layoutManager = LinearLayoutManager(activity)
        binding.repoIssuesRecyclerView.addItemDecoration(DividerItemDecoration(activity, DividerItemDecoration.VERTICAL))
    }

    private fun initSwipeRefreshLayout() {
        binding.repoIssuesSwipeRefreshLayout.setOnRefreshListener {
            repoIssuesCall(authHeader, userName, repoName)
            binding.repoIssuesSwipeRefreshLayout.isRefreshing = false
        }
    }

    private fun repoIssuesCall(header: String, user: String, repo: String) {
        disposables.add(viewModel.loadRepoIssues(header, user, repo)
                .subscribe({ repoIssueList ->
                    if (repoIssueList.isNotEmpty()) {
                        val issueAdapter = IssueAdapter(repoIssueList)
                        binding.repoIssuesRecyclerView.adapter = issueAdapter
                    } else {
                        binding.noIssuesTextView.visibility = View.VISIBLE
                    }
                }, { error -> Timber.e(error.message) })
        )
    }

}
